### PR TITLE
SNAPSYNC add request task

### PR DESCRIPTION
**Pull request by matkt**
## PR description

PR that adds the different request tasks necessary for the snapsync as well as a utility to manage the ranges of requests 

Snapsyc is still an experimental feature more tests will be added when the feature will be completed

## Changelog

- [ ] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/messages/snap/AccountRangeMessageTest.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright ConsenSys AG.
```
Comment (fab-10) [inline ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/messages/snap/AccountRangeMessageTest.java:2]: Update the copyright.
This comment applies also to other classes in this package.

File: ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/RangeManagerTest.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        RangeManager.generateRanges(
+            Bytes32.fromHexString(
+                "0x0000000000000000000000000000000000000000000000000000000000000000"),
+            Bytes32.fromHexString(
+                "0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"),
+            3);
+    Assertions.assertThat(ranges.size()).isEqualTo(3);
+    Assertions.assertThat(ranges).isEqualTo(expectedResult);
+  }
+}
```
Comment (fab-10) [inline ethereum/eth/src/test/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/RangeManagerTest.java:83]: Unit tests for `findNewBeginElementInRange` methods are missing

**Review by fab-10 — CHANGES_REQUESTED** [review]

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/messages/snap/GetByteCodesMessage.java [hunk 3]
```diff
@@ -41,40 +40,31 @@ public static GetByteCodesMessage readFrom(final MessageData message) {
...
     tmp.endList();
-    return new GetByteCodesMessage(accountHashes, tmp.encoded());
+    return new GetByteCodesMessage(tmp.encoded());
   }
 
-  public GetByteCodesMessage(final Optional<ArrayDeque<Bytes32>> accountHashes, final Bytes data) {
+  public GetByteCodesMessage(final Bytes data) {
     super(data);
-    this.accountHashes = accountHashes;
   }
```
[the remark this answers is not in the document]
Comment (matkt) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/messages/snap/GetByteCodesMessage.java]: done

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/RangeManager.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+   */
+  public static Optional<Bytes32> findNewBeginElementInRange(
+      final Bytes32 worldstateRootHash,
+      final List<Bytes> proofs,
+      final TreeMap<Bytes32, Bytes> receivedKeys,
+      final Bytes32 endKeyHash) {
+    if (receivedKeys.isEmpty() || receivedKeys.lastKey().compareTo(endKeyHash) >= 0) {
+      return Optional.empty();
+    } else {
+      final Map<Bytes32, Bytes> proofsEntries = Collections.synchronizedMap(new HashMap<>());
```
[the remark this answers is not in the document]
Comment (matkt) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/RangeManager.java]: yes good catch it was link to another version of the code

[the remark this answers is not in the document]
Comment (matkt) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/snapsync/RangeManager.java]: done

**Review by fab-10 — APPROVED** [review]

**Pull request merged**